examples/custom_storage.ml

```diff
@@ -0,0 +1,114 @@
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Lwt.Syntax
+
+module Hashtbl_config = struct
+  module Conf = Irmin.Backend.Conf
+
+  let spec = Conf.Spec.v "hashtbl"
+  let init_size = Conf.key ~spec "init-size" Irmin.Type.int 8
+  let empty = Conf.empty spec
+end
+
+module Hashtbl_storage : Irmin.Simple_storage.Maker =
+functor
+  (Key : Irmin.Type.S)
+  (Value : Irmin.Type.S)
+  ->
+  struct
+    module Tbl = Hashtbl.Make (struct
+      type t = Key.t
+
+      let equal a b = Irmin.Type.(unstage (equal Key.t)) a b
+      let hash k = Irmin.Type.(unstage (short_hash Key.t)) k
+    end)
+
+    (** Types *)
+
+    type 'a t = Value.t Tbl.t
+    type key = Key.t
+    type value = Value.t
+
+    (** Initialisation / Closing *)
+
+    let v config =
+      let init_size = Irmin.Backend.Conf.get config Hashtbl_config.init_size in
+      Tbl.create init_size |> Lwt.return
+
+    let close _t = Lwt.return_unit
+
+    (** Operations *)
+
+    let set t key value = Tbl.replace t key value |> Lwt.return
+    let mem t key = Tbl.mem t key |> Lwt.return
+    let find t key = Tbl.find_opt t key |> Lwt.return
+    let keys t = Tbl.to_seq_keys t |> List.of_seq |> Lwt.return
+    let remove t key = Tbl.remove t key |> Lwt.return
+    let clear t = Tbl.clear t |> Lwt.return
+    let cast t = (t :> Irmin.Perms.read_write t)
+    let lock = Mutex.create ()
+
+    let batch t f =
+      Mutex.lock lock;
+      let+ x =
+        Lwt.catch
+          (fun () -> f (cast t))
+          (fun exn ->
+            Mutex.unlock lock;
+            raise exn)
+      in
+      Mutex.unlock lock;
+      x
+  end
+
+module Content_addressable =
+  Irmin.Simple_storage.Content_addressable (Hashtbl_storage)
+
+module Atomic_write = Irmin.Simple_storage.Atomic_write (Hashtbl_storage)
+
+module KV = struct
+  module Maker = Irmin.Maker (Content_addressable) (Atomic_write)
+
+  module Make (C : Irmin.Contents.S) = struct
+    include Maker.Make (struct
+      module Info = Irmin.Info.Default
+      module Metadata = Irmin.Metadata.None
+      module Contents = C
+      module Path = Irmin.Path.String_list
+      module Branch = Irmin.Branch.String
+      module Hash = Irmin.Hash.SHA1
+      module Node = Irmin.Node.Make (Hash) (Path) (Metadata)
+      module Commit = Irmin.Commit.Make (Hash)
+    end)
+  end
+end
+
+module Store = KV.Make (Irmin.Contents.String)
```

Comment:
I feel a need to balance between having a one-line default config Store and demonstrating how to configure a store. Thinking a little more based on your comment, I like the idea of adding something like 

`module Store = Irmin.Simple_storage.Store (Hashtble_storage) (Irmin.Hash.SHA1) (Irmin.Contents.String)` 

since hash is the most likely candidate for configuration and already goes with the "maker" pattern of key and value as parameters.

Reply:
This is now `module Store = Irmin.Of_simple_storage (Hashtble_storage) (Irmin.Hash.SHA1) (Irmin.Contents.String)`. Thanks for the nudge!

Reply:
Way better!